src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -0,0 +1,129 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CoreService } from 'src/core/types';
+import { take, retryWhen, concatMap } from 'rxjs/operators';
+import { defer, throwError, iif, timer } from 'rxjs';
+import elasticsearch from 'elasticsearch';
+import { KibanaMigrator } from './migrations';
+import { CoreContext } from '../core_context';
+import { LegacyServiceSetup } from '../legacy/legacy_service';
+import { ElasticsearchServiceSetup, CallAPIOptions } from '../elasticsearch';
+import { KibanaConfig } from '../kibana_config';
+
+/**
+ * @public
+ */
+export interface SavedObjectsServiceStart {
+  migrator: Pick<KibanaMigrator, keyof KibanaMigrator>;
+}
+
+/**
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SavedObjectsServiceSetup {}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface StartDeps {}
+
+export interface SetupDeps {
+  legacy: LegacyServiceSetup;
+  elasticsearch: ElasticsearchServiceSetup;
+}
+
+export class SavedObjectsService
+  implements CoreService<SavedObjectsServiceSetup, SavedObjectsServiceStart> {
+  private migrator: KibanaMigrator | undefined;
+
+  constructor(private readonly coreContext: CoreContext) {}
+
+  public async setup(coreSetup: SetupDeps) {
+    const {
+      savedObjectSchemas,
+      savedObjectMappings,
+      savedObjectMigrations,
+      savedObjectValidations,
+    } = await coreSetup.legacy.uiExports;
+
+    const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(take(1)).toPromise();
+
+    const retryCallCluster = (
+      endpoint: string,
+      clientParams: Record<string, any> = {},
+      options?: CallAPIOptions
+    ) => {
+      return defer(() => adminClient.callAsInternalUser(endpoint, clientParams, options))
+        .pipe(
+          retryWhen(errors =>
+            errors.pipe(
+              concatMap((error, i) =>
+                iif(
+                  () => error instanceof elasticsearch.errors.NoConnections,
+                  timer(1000),
+                  throwError(error)
+                )
+              )
+            )
+          )
+        )
+        .toPromise();
+    };
+
+    // const kibanaConfig = await this.coreContext.configService
+    //   .atPath<KibanaConfig>('kibana')
+    //   .toPromise(); <-- No idea why, but this makes Node.js process exit without any errors or hints to what went wrong ??
+    let kibanaConfig;
+    this.coreContext.configService.atPath<KibanaConfig>('kibana').subscribe(value => {
+      kibanaConfig = value;
+    });
```

Comment:
Thanks the `pipe(take(1))` cleans up the code. I managed to reproduce it by creating a new ReplaySubject(1) and then awaiting `.toPromise()` without a `.pipe(take(1))`. I would expect it to just never finish the await since the replay subject won't close, but it crashes :/